[WebAssembly] Add patterns for relaxed madd

badumbatish opened the pull request:
[WebAssembly] Fold fadd contract (fmul contract) to relaxed madd w/ -mattr=+simd128,+relaxed-simd

Fixes #121311

- Precommit test for #121311
- Fold fadd contract (fmul contract) to relaxed madd w/ -mattr=+simd128,+relaxed-simd
- Move PatFrag of fadd_contract in ARM.td and WebAssembly.td to TargetSelectionDAG.td for reuse of pattern

https://github.com/llvm/llvm-project/issues/55932 tracks this

badumbatish committed:
Precommit test for #121311

badumbatish committed:
[WASM] Optimize fma when relaxed and ffast-math

Fixes #121311, which folds a series of multiply and add to wasm.fma when
we have -mrelaxed-simd and -ffast-math.

Also attempted to use wasm.fma instead of the built in llvm.fma

arsenm commented on llvm/lib/Target/WebAssembly/WebAssemblyISelLowering.cpp:
```diff
@@ -3412,6 +3418,37 @@ static SDValue performSETCCCombine(SDNode *N,
+static SDValue performFAddCombine(SDNode *N, SelectionDAG &DAG) {
+  assert(N->getOpcode() == ISD::FADD);
+  using namespace llvm::SDPatternMatch;
+  if (!N->getFlags().hasFastMath())
+    return SDValue();
```
This does not require all fast math flags

arsenm commented on llvm/include/llvm/CodeGen/SelectionDAGNodes.h:
```diff
@@ -475,6 +475,7 @@ struct SDNodeFlags {
   bool hasAllowReassociation() const { return Flags & AllowReassociation; }
   bool hasNoFPExcept() const { return Flags & NoFPExcept; }
   bool hasUnpredictable() const { return Flags & Unpredictable; }
+  bool hasFastMath() const { return Flags & FastMathFlags; }
```
Should not add this. This does not depend on all the flags, and you should only depend on the exact set of flags required for a transform

arsenm commented on llvm/test/CodeGen/WebAssembly/simd-relaxed-fma.ll:
```diff
@@ -0,0 +1,43 @@
+; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py UTC_ARGS: --version 5
+
+; RUN: llc < %s -verify-machineinstrs -disable-wasm-fallthrough-return-opt -wasm-disable-explicit-locals -wasm-keep-registers  -mattr=+simd128,+relaxed-simd | FileCheck %s
```
Don't need -verify-machineinstrs

arsenm commented on llvm/lib/Target/WebAssembly/WebAssemblyISelLowering.cpp:
```diff
@@ -3412,6 +3418,37 @@ static SDValue performSETCCCombine(SDNode *N,
+  return SDValue();
+}
+
+static SDValue performFMACombine(SDNode *N, SelectionDAG &DAG) {
+  assert(N->getOpcode() == ISD::FMA);
+  if (!N->getFlags().hasFastMath())
+    return SDValue();
+
+  SDLoc DL(N);
+  SDValue A = N->getOperand(0), B = N->getOperand(1), C = N->getOperand(2);
+  EVT VecVT = N->getValueType(0);
+
+  return DAG.getNode(
+      ISD::INTRINSIC_WO_CHAIN, DL, VecVT,
+      {DAG.getConstant(Intrinsic::wasm_relaxed_madd, DL, MVT::i32), A, B, C});
+}
```
I doubt it's appropriate to turn an FMA into a not-FMA, regardless of fast math flags. What are the semantics of wasm_relaxed_madd?

arsenm replied:
Which is not FMA, this is equivalent to llvm.fmuladd. I would just use llvm.fmuladd and not introduce this custom intrinsic

badumbatish replied:
got it, thanks Matt, will move fadd fmul -> relaxed madd to tablegen

arsenm commented on llvm/test/CodeGen/WebAssembly/simd-relaxed-fma.ll:
```diff
@@ -0,0 +1,43 @@
+; CHECK-NEXT:    v128.store 0($3):p2align=0, $pop3
+; CHECK-NEXT:    return
+entry:
+  %0 = load <4 x float>, ptr %a, align 1
+  %1 = load <4 x float>, ptr %b, align 1
+  %2 = load <4 x float>, ptr %c, align 1
```
Better to use argument values instead of loading the sample values

arsenm commented on llvm/test/CodeGen/WebAssembly/simd-relaxed-fma.ll:
```diff
@@ -0,0 +1,43 @@
+  %fma = tail call fast <4 x float> @llvm.fma.v4f32(<4 x float> %0, <4 x float> %1, <4 x float> %2)
+  store <4 x float> %fma, ptr %dest, align 1
+  ret void
+}
```
Test scalar cases and other vector widths

badumbatish committed:
[WASM] Fix nits for PR 147487

- Fix inefficient wasm test case.
- Added scalar test case and more floating type.
- Remove total ffast checking -> allowContract

badumbatish committed:
[WASM] Add more vector widths for PR 147487

- Added support for <8 x f32>.
- Refactored out condition for relaxed simd to a seperate function.

badumbatish commented:
@lukel97 hi Luke! can I have some comments on the PR? ty!

lukel97 reviewed (commented):
These combines look straightforward enough, can they be implemented as patterns in tablegen?

lukel97 commented on llvm/lib/Target/WebAssembly/WebAssemblyISelLowering.cpp:
```diff
@@ -3411,6 +3418,62 @@ static SDValue performSETCCCombine(SDNode *N,
+  SDLoc DL(N);
+  SDValue A, B, C;
+  SelectionDAG &DAG = DCI.DAG;
+  if (sd_match(N, m_FAdd(m_Value(A), m_FMul(m_Value(B), m_Value(C)))))
```
It's probably best to not try and combine fma nodes ourselves, since I think we also need to check the fast math flags on the multiply too. 

DAGCombiner or InstCombine or something probably already does something similar when supported. Can we reduce this PR to just handle the llvm.fma intrinsic to begin with?

arsenm replied:
There is quite a lot of FMA forming logic you shouldn't try to reinvent in a backend

arsenm commented on llvm/lib/Target/WebAssembly/WebAssemblyISelLowering.cpp:
```diff
@@ -3411,6 +3418,62 @@ static SDValue performSETCCCombine(SDNode *N,
+  SelectionDAG &DAG = DCI.DAG;
+  if (sd_match(N, m_FAdd(m_Value(A), m_FMul(m_Value(B), m_Value(C)))))
+    return DAG.getNode(
+        ISD::INTRINSIC_WO_CHAIN, DL, VecVT,
+        {DAG.getConstant(Intrinsic::wasm_relaxed_madd, DL, MVT::i32), A, B, C});
```
It seems this is equivalent to ISD::FMAD? Report that as legal and use it instead?

lukel97 replied:
wasm's madd is might either be rounded once or twice, depending on the implementation. Just from reading the comment here it seems like FMAD requires exactly two instances of rounding?

```
 /// FMAD - Perform a * b + c, while getting the same result as the
 /// separately rounded operations.
```

I think the precise equivalent is probably something like `fadd x (fmul contract y z)`, where it's not specified if it needs to be rounded once or twice?

badumbatish replied:
it does look like llvm.fmuladd requires `fast` flags with it, but its currently a TODO in the codebase

Initial scouting:

- Part where it requires fast math:
```tablegen
class LLVM_TernarySameArgsIntrOpF<string func, list<Trait> traits = []> :
    LLVM_TernarySameArgsIntrOpBase<func, LLVM_AnyFloat, traits,
                                  /*requiresFastmath=*/1> {
  dag fmfArg = (
    ins DefaultValuedAttr<LLVM_FastmathFlagsAttr, "{}">:$fastmathFlags);
  let arguments = !con(commonArgs, fmfArg);
}
...
def LLVM_FMulAddOp : LLVM_TernarySameArgsIntrOpF<"fmuladd">;
```

- Part where it wants fast math propagation?!:
```c++
// /llvm/lib/CodeGen/SelectionDAG/SelectionDAGBuilder.cpp:6995
  case Intrinsic::fmuladd: {
    EVT VT = TLI.getValueType(DAG.getDataLayout(), I.getType());
    if (TM.Options.AllowFPOpFusion != FPOpFusion::Strict &&
        TLI.isFMAFasterThanFMulAndFAdd(DAG.getMachineFunction(), VT)) {
      setValue(&I, DAG.getNode(ISD::FMA, sdl,
                               getValue(I.getArgOperand(0)).getValueType(),
                               getValue(I.getArgOperand(0)),
                               getValue(I.getArgOperand(1)),
                               getValue(I.getArgOperand(2)), Flags));
    } else {
      // TODO: Intrinsic calls should have fast-math-flags.
      SDValue Mul = DAG.getNode(
          ISD::FMUL, sdl, getValue(I.getArgOperand(0)).getValueType(),
          getValue(I.getArgOperand(0)), getValue(I.getArgOperand(1)), Flags);
      SDValue Add = DAG.getNode(ISD::FADD, sdl,
                                getValue(I.getArgOperand(0)).getValueType(),
                                Mul, getValue(I.getArgOperand(2)), Flags);
      setValue(&I, Add);
    }
    return;
  }

badumbatish replied:
; RUN: llc < %s -disable-wasm-fallthrough-return-opt -wasm-disable-explicit-locals -wasm-keep-registers  -mattr=+simd128,+relaxed-simd | FileCheck %s --check-prefix=RELAXED
; RUN: llc < %s -disable-wasm-fallthrough-return-opt -wasm-disable-explicit-locals -wasm-keep-registers  -mattr=+simd128,              | FileCheck %s --check-prefix=STRICT


this is if I don't pass in the -fast flag in the call
```
define <4 x float> @fmuladd_vector_4xf32(<4 x float> %a, <4 x float> %b, <4 x float> %c) {
; RELAXED-LABEL: fmuladd_vector_4xf32:
; RELAXED:         .functype fmuladd_vector_4xf32 (v128, v128, v128) -> (v128)
; RELAXED-NEXT:  # %bb.0: # %entry
; RELAXED-NEXT:    f32x4.mul $push0=, $0, $1
; RELAXED-NEXT:    f32x4.add $push1=, $pop0, $2
; RELAXED-NEXT:    return $pop1
;
; STRICT-LABEL: fmuladd_vector_4xf32:
; STRICT:         .functype fmuladd_vector_4xf32 (v128, v128, v128) -> (v128)
; STRICT-NEXT:  # %bb.0: # %entry
; STRICT-NEXT:    f32x4.mul $push0=, $0, $1
; STRICT-NEXT:    f32x4.add $push1=, $pop0, $2
; STRICT-NEXT:    return $pop1
entry:
  %fma = tail call <4 x float> @llvm.fmuladd(<4 x float> %a, <4 x float> %b, <4 x float> %c)
  ret <4 x float> %fma
}
```

badumbatish replied:
this is if i do
```
define <4 x float> @fmuladd_vector_4xf32(<4 x float> %a, <4 x float> %b, <4 x float> %c) {
; RELAXED-LABEL: fmuladd_vector_4xf32:
; RELAXED:         .functype fmuladd_vector_4xf32 (v128, v128, v128) -> (v128)
; RELAXED-NEXT:  # %bb.0: # %entry
; RELAXED-NEXT:    f32x4.relaxed_madd $push0=, $2, $0, $1
; RELAXED-NEXT:    return $pop0
;
; STRICT-LABEL: fmuladd_vector_4xf32:
; STRICT:         .functype fmuladd_vector_4xf32 (v128, v128, v128) -> (v128)
; STRICT-NEXT:  # %bb.0: # %entry
; STRICT-NEXT:    f32x4.mul $push0=, $0, $1
; STRICT-NEXT:    f32x4.add $push1=, $pop0, $2
; STRICT-NEXT:    return $pop1
entry:
  %fma = tail call fast <4 x float> @llvm.fmuladd(<4 x float> %a, <4 x float> %b, <4 x float> %c)
  ret <4 x float> %fma
}
```

badumbatish replied:
I added test cases for both just in case

arsenm replied:
fmuladd has nothing to do with fast math flags. I don't know what the MLIR thing is on about, but it's nonsense. fmuladd corresponds to the C behavior where single expression containing multiply and add may contract.

The emission time preference as the split ops or FMA is a target preference, it is not dependent on fast flags

badumbatish replied:
hmm, ok, @arsenm should I make fmuladd legal now in the webassembly backend, and match it to produce relaxed madd, or should I fix the todo `// TODO: Intrinsic calls should have fast-math-flags.` by adding `contract` flag and rely on fadd fmul to produce relaxed madd?

badumbatish committed:
Use tablegen instead of cpp, relax only seperate fmul fadd

- Use tablegen instead of cpp
- Relax only seperate fmul fadd instead of fma.
- Fix test name, add negative testing, add fmuladd tests, both fast and
  nonfast

lukel97 commented on llvm/test/CodeGen/WebAssembly/simd-relaxed-fma.ll:
```diff
@@ -0,0 +1,276 @@
+; STRICT-NEXT:    f32x4.add $push1=, $pop0, $2
+; STRICT-NEXT:    return $pop1
+entry:
+  %mul.i = fmul fast <4 x float> %b, %a
+  %add.i = fadd fast <4 x float> %mul.i, %c
```
I think we only need the contract flag, not the entire set of fast flags
```suggestion
  %mul.i = fmul contract <4 x float> %b, %a
  %add.i = fadd contract <4 x float> %mul.i, %c
```

lukel97 commented on llvm/lib/Target/WebAssembly/WebAssemblyInstrSIMD.td:
```diff
@@ -1542,6 +1542,29 @@ defm "" : SIMDMADD<F32x4, 0x105, 0x106, [HasRelaxedSIMD]>;
+  return N->getFlags().hasAllowContract();
+}]>;
+
+def : Pat<(fadd (v4f32 V128:$a), (fmul_fast (v4f32 V128:$b), (v4f32 V128:$c))),
```
Based on this small test, dag combiner only seems to fuse this to an fma when both nodes have the contract flag set:

```llvm
define float @f(float %x, float %y, float %z) {
  %1 = fmul contract float %x, %y
  %2 = fadd contract float %1, %z
  ret float %2
}
```

So we should probably also restrict our pattern to when both nodes have contract too

lukel97 commented on llvm/lib/Target/WebAssembly/WebAssemblyInstrSIMD.td:
```diff
@@ -1542,6 +1542,29 @@ defm "" : SIMDMADD<F32x4, 0x105, 0x106, [HasRelaxedSIMD]>;
 defm "" : SIMDMADD<F64x2, 0x107, 0x108, [HasRelaxedSIMD]>;
 defm "" : SIMDMADD<F16x8, 0x14e, 0x14f, [HasFP16]>;
 
+def fmul_fast : PatFrag<(ops node:$a, node:$b), (fmul node:$a, node:$b),[{
```
```suggestion
def fmul_contract : PatFrag<(ops node:$a, node:$b), (fmul node:$a, node:$b),[{
```

lukel97 replied:
Actually it looks like ARMInstrInfo.td also defines fadd_contract. Maybe we should move that definition  into include/llvm/Target/TargetSelectionDAG.td + copy it for fmul_contract, that way we can share it between targets?

badumbatish replied:
done

lukel97 commented:
PR title + description looks like it needs updated, the title should also have the `[WebAssembly]` tag

badumbatish committed:
Resolve PR review

- Move PatFrag from ARM.td and WebAssembly.td to TargetSelectionDAG.td
- Rename and reformat tests to a more consistent pattern.
- Added todos for test.

lukel97 commented on llvm/test/CodeGen/WebAssembly/simd-relaxed-fma.ll:
```diff
@@ -0,0 +1,249 @@
+; STRICT-NEXT:    call $push18=, fmaf, $pop17, $pop16, $pop15
+; STRICT-NEXT:    f32x4.replace_lane $push19=, $pop14, 3, $pop18
+; STRICT-NEXT:    return $pop19
+  %fma = tail call contract <4 x float> @llvm.fma(<4 x float> %a, <4 x float> %b, <4 x float> %c)
```
No need for the contract flag here since by definition fma happens in one rounding step

```suggestion
  %fma = call <4 x float> @llvm.fma(<4 x float> %a, <4 x float> %b, <4 x float> %c)
```

lukel97 commented on llvm/test/CodeGen/WebAssembly/simd-relaxed-fma.ll:
```diff
@@ -0,0 +1,249 @@
+; STRICT-NEXT:  # %bb.0:
+; STRICT-NEXT:    call $push0=, fma, $0, $1, $2
+; STRICT-NEXT:    return $pop0
+  %fma = tail call contract double @llvm.fma(double %a, double %b, double %c)
```
```suggestion
  %fma = call double @llvm.fma(double %a, double %b, double %c)
```

lukel97 commented on llvm/test/CodeGen/WebAssembly/simd-relaxed-fma.ll:
```diff
@@ -0,0 +1,249 @@
+  ret <4 x float> %fma
+}
+
+; TODO: This should also have relaxed_madd in RELAXED case
```
Yeah I think leaving a TODO is the best thing to do for now. We're losing the information in SelectionDAGBuilder that a fmuladd can be either fused or not. 

After thinking about it a bit, I don't think we can just slap on contract flags either since that allows other types of contraction. We might need a new node or something but this is something to revisit in the future

lukel97 commented on llvm/lib/Target/WebAssembly/WebAssemblyInstrSIMD.td:
```diff
@@ -1542,6 +1542,13 @@ defm "" : SIMDMADD<F32x4, 0x105, 0x106, [HasRelaxedSIMD]>;
 defm "" : SIMDMADD<F64x2, 0x107, 0x108, [HasRelaxedSIMD]>;
 defm "" : SIMDMADD<F16x8, 0x14e, 0x14f, [HasFP16]>;
 
+def : Pat<(fadd_contract (v4f32 V128:$a), (fmul_contract (v4f32 V128:$b), (v4f32 V128:$c))),
+           (MADD_F32x4 V128:$a, V128:$b, V128:$c)>, Requires<[HasRelaxedSIMD]>;
```
Not a blocker, but is it possible to move this pattern into the SIMDMADD multiclass so we don't need to redefine it for each type?

E.g. can we add it to the list of patterns here somehow

```
[(set (vec.vt V128:$dst), (int_wasm_relaxed_madd
             (vec.vt V128:$a), (vec.vt V128:$b), (vec.vt V128:$c))),
 (set (vec.vt V128:$dst), (fadd_contract (v4f32 V128:$a),
             (fmul_contract (v4f32 V128:$b), (v4f32 V128:$c)))]
```

I think that way we'll also get fp16 for free

badumbatish replied:
i tried it out but the tablegen gives an error on `cannot define dst`. I'm unsure if its the limit of tablegen or is there sth can be done.

I can add a pattern like this, which also resides in the multiclass
```diff
multiclass SIMDMADD<Vec vec, bits<32> simdopA, bits<32> simdopS, list<Predicate> reqs> {
  defm MADD_#vec :
    SIMD_I<(outs V128:$dst), (ins V128:$a, V128:$b, V128:$c), (outs), (ins),
           [(set (vec.vt V128:$dst), (int_wasm_relaxed_madd
             (vec.vt V128:$a), (vec.vt V128:$b), (vec.vt V128:$c)))],
           vec.prefix#".relaxed_madd\t$dst, $a, $b, $c",
           vec.prefix#".relaxed_madd", simdopA, reqs>;
  defm NMADD_#vec :
    SIMD_I<(outs V128:$dst), (ins V128:$a, V128:$b, V128:$c), (outs), (ins),
           [(set (vec.vt V128:$dst), (int_wasm_relaxed_nmadd
             (vec.vt V128:$a), (vec.vt V128:$b), (vec.vt V128:$c)))],
           vec.prefix#".relaxed_nmadd\t$dst, $a, $b, $c",
           vec.prefix#".relaxed_nmadd", simdopS, reqs>;

+  def : Pat<(fadd_contract (vec.vt V128:$a), (fmul_contract (vec.vt V128:$b), (vec.vt V128:$c))),
+            (!cast<Instruction>("MADD_"#vec) V128:$a, V128:$b, V128:$c)>, Requires<[HasRelaxedSIMD]>;

}
```

lukel97 reviewed (approved):
Generally LGTM, thanks

badumbatish committed:
Added test case for fp16, refactor pattern support in tablegen to multiclass

lukel97 commented on llvm/lib/Target/WebAssembly/WebAssemblyInstrSIMD.td:
```diff
@@ -1536,6 +1536,10 @@ multiclass SIMDMADD<Vec vec, bits<32> simdopA, bits<32> simdopS, list<Predicate>
            vec.prefix#".relaxed_nmadd", simdopS, reqs>;
+
+  def : Pat<(fadd_contract (vec.vt V128:$a), (fmul_contract (vec.vt V128:$b), (vec.vt V128:$c))),
+             (!cast<Instruction>("MADD_"#vec) V128:$a, V128:$b, V128:$c)>, Requires<[HasRelaxedSIMD]>;
```
I think you can just move this inside the actual MADD definition? Alongside where the pattern for the intrinsic is defined

lukel97 replied:
Yup my bad! I think this Pat is plenty fine

lukel97 reviewed (approved):
LGTM but I'd wait for a review from the wasm folks first

lukel97 commented:
I've gone ahead and committed this. Thanks again @dschuff for helping review a lot of these patches, it's much appreciated!
